src/App/Fossa/LicenseScanner.hs

```diff
@@ -0,0 +1,119 @@
+
+runThemis :: (Has Exec sig m, Has Diagnostics sig m) => ThemisBins -> Path Abs Dir -> m [LicenseUnit]
+runThemis themisBins scanDir = do
+  context "Running license scan binary" $ execThemis themisBins scanDir
+
+md5 :: B.ByteString -> Digest MD5
+md5 = hash
+
+scanAndUploadVendoredDep :: (Has Diag.Diagnostics sig m, Has (Lift IO) sig m, Has StickyLogger sig m) => ApiOpts -> Path Abs Dir -> VendoredDependency -> m Archive
+scanAndUploadVendoredDep apiOpts baseDir VendoredDependency{..} = context "compressing and uploading vendored deps" $ do
+  logSticky $ "License Scanning '" <> vendoredName <> "' at '" <> vendoredPath <> "'"
+  vendoredDepDir <- case parseRelDir $ toString vendoredPath of
+    Left _ -> fatalText "Error constructing scan dir for vendored scan"
+    Right val -> pure val
+  let cliScanDir = baseDir </> vendoredDepDir
+  themisScanResult <- runLicenseScanOnDir cliScanDir
+  let licenseSourceUnit =
+        LicenseSourceUnit
+          { licenseSourceUnitName = vendoredPath
+          , licenseSourceUnitType = "cli-license-scanned"
+          , licenseSourceUnitLicenseUnits = themisScanResult
+          }
+
+  depVersion <- case vendoredVersion of
+    Nothing -> pure $ toText $ show $ md5 $ encodeUtf8 $ show themisScanResult
+    Just version -> pure version
+
+  signedURL <- Fossa.getSignedLicenseScanURL apiOpts depVersion vendoredName
+
+  logSticky $ "Uploading '" <> vendoredName <> "' to secure S3 bucket"
+  _ <- Fossa.licenseScanResultUpload signedURL licenseSourceUnit
+
+  pure $ Archive vendoredName depVersion
+
+-- archiveUploadSourceUnit receives a list of vendored dependencies, a root path, and API settings.
+-- Using this information, it license scans each vendored dependency, uploads the license scan results and then queues a build for the dependency.
+licenseScanSourceUnit :: (Has Diag.Diagnostics sig m, Has (Lift IO) sig m, Has StickyLogger sig m) => Path Abs Dir -> ApiOpts -> [VendoredDependency] -> m [Locator]
+licenseScanSourceUnit baseDir apiOpts vendoredDeps = do
+  -- Users with many instances of vendored dependencies may accidentally have complete duplicates. Remove them.
+  let uniqDeps = nub vendoredDeps
+
+  -- However, users may also have vendored dependencies that have duplicate names but are not complete duplicates.
+  -- These aren't valid and can't be automatically handled, so fail the scan with them.
+  let duplicates = duplicateNames uniqDeps
+  unless (null duplicates) $ Diag.fatalText $ duplicateFailureBundle duplicates
+
+  -- At this point, we have a good list of deps, so go for it.
+  archives <- traverse (scanAndUploadVendoredDep apiOpts baseDir) uniqDeps
+
+  -- archiveBuildUpload takes archives without Organization information. This orgID is appended when creating the build on the backend.
+  -- We don't care about the response here because if the build has already been queued, we get a 401 response.
+  _ <- Fossa.licenseScanFinalize apiOpts $ ArchiveComponents archives
```

Comment:
The build is queued by this call, so either

1. The function returns non-fatal because the build was successfully queued by this call.
2. The function returns non-fatal because the build was already queued, and we ignored the 401 error (legacy core issue) that signals that.
3. The function returns fatal because something went wrong trying to queue the build.